src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java

```diff
@@ -0,0 +1,190 @@
+import java.util.concurrent.Phaser;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.mskcc.cmo.messaging.Gateway;
+import org.mskcc.cmo.messaging.MessageConsumer;
+import org.mskcc.cmo.metadb.model.SampleMetadata;
+import org.mskcc.cmo.metadb.service.CmoLabelGeneratorService;
+import org.mskcc.cmo.metadb.service.RequestReplyHandlingService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+/**
+ *
+ * @author DivyaMadala
+ *
+ */
+
+@Service
+public class RequestReplyHandlingServiceImpl implements RequestReplyHandlingService {
+
+    @Value("${request_reply.cmo_label_generator_topic}")
+    private String CMO_LABEL_GENERATOR_REQREPLY_TOPIC;
+
+    @Value("${request_reply.patient_samples_topic}")
+    private String PATIENT_SAMPLES_REQUEST_TOPIC;
+
+    @Value("${num.new_request_handler_threads}")
+    private int NUM_NEW_REQUEST_HANDLERS;
+
+    @Autowired
+    private CmoLabelGeneratorService cmoLabelGeneratorService;
+
+    private final ObjectMapper mapper = new ObjectMapper();
+    private static Gateway messagingGateway;
+    private static final Log LOG = LogFactory.getLog(RequestReplyHandlingServiceImpl.class);
+    private static boolean initialized = false;
+    private static volatile boolean shutdownInitiated;
+
+    private static final ExecutorService exec = Executors.newCachedThreadPool();
+    private static final BlockingQueue<ReplyInfo> cmoLabelGeneratorReqReplyQueue =
+            new LinkedBlockingQueue<ReplyInfo>();
+    private static CountDownLatch cmoLabelGeneratorHandlerShutdownLatch;
+
+
+    private class ReplyInfo {
+        String requestMessage;
+        String replyTo;
+
+        ReplyInfo(String requestMessage, String replyTo) {
+            this.requestMessage = requestMessage;
+            this.replyTo = replyTo;
+        }
+
+        String getRequestMessage() {
+            return requestMessage;
+        }
+
+        String getReplyTo() {
+            return replyTo;
+        }
+    }
+
+    private class CmoLabelGeneratorReqReplyHandler implements Runnable {
+
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        CmoLabelGeneratorReqReplyHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    // reply info request message contains cmo patient id
+                    ReplyInfo replyInfo = cmoLabelGeneratorReqReplyQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if (replyInfo != null) {
+                        SampleMetadata sample = mapper.readValue(replyInfo.getRequestMessage(),
+                                SampleMetadata.class);
+                        String updatedCmoSampleLabel =
```

Comment:
I think we want to look at the list of existing samples to check if any of them match the igoId/primaryId of the incoming sample we received

Reply:
more details here https://app.zenhub.com/workspaces/metadb-scrum-5fcfa28a4ccc1600165347b6/issues/mskcc/cmo-metadb/598